pkg/segment/search/searchaggs.go

```diff
@@ -254,9 +256,20 @@ func addRecordToAggregations(grpReq *structs.GroupByRequest, timeHistogram *stru
 				if err != nil {
 					log.Errorf("addRecordToAggregations: Failed to get key for column %v: %v", byField, err)
 				} else {
-					strs, err := utils.ConvertGroupByKey(rawVal)
-					if err != nil {
-						log.Errorf("addRecordToAggregations: failed to extract raw key: %v", err)
+					rawValStr := toputils.UnsafeByteSliceToString(rawVal) // Zero copy, if we get a cache hit.
+					strs, exists := groupByCache[rawValStr]
+					if !exists {
+						strs, err = utils.ConvertGroupByKey(rawVal)
+						if err != nil {
+							log.Errorf("addRecordToAggregations: failed to extract raw key: %v", err)
+						}
+
+						// I'm pretty sure we need to actually copy the string
+						// here to insert it into the map, since we made the
+						// string previously with an unsafe conversion from a
+						// []byte, and that []byte will change later.
+						rawValStr = string(rawVal)
+						groupByCache[rawValStr] = strs
```

Comment:
if we have failed to convertGroupBY.. then do we still want to insert it into the cache ?

Reply:
Good question. I'm actually not sure what we should do. Currently `ConvertGroupByKey` is deterministic and doesn't rely on any external state, so if it fails for a certain input, it will always fail on that input. So caching here is safe for now, but may change if `ConvertGroupByKey` changes

In general though maybe we wouldn't want to cache results that return an error because it could have errored due to some external state. What do you think we should do here?

Reply:
for now we should not store it in cache